File: baekjoon/BinarySearch/10816_숫자 카드2.js
const fs = require('fs')
let [n, nArr, m, mArr] = fs
  .readFileSync('../input.txt')
  .toString()
  .trim()
  .split('\n')
n = Number(n)
nArr = nArr
  .split(' ')
  .map(Number)
  .sort((a, b) => a - b)
mArr = mArr.split(' ').map(Number)

function solution() {
  const dic = {}
  // 중복 요소 개수 카운팅
  nArr.forEach((x) => {
    dic[x] = (dic[x] || 0) + 1
  })
  // 중복 제거
  const arr = [...new Set(nArr)]

  const answer = []

  for (const find of mArr) {
    let start = 0
    let end = arr.length - 1
    let check = false
    let mid = 0
    while (start <= end) {
      mid = parseInt((start + end) / 2)
      if (arr[mid] === find) {
        check = true
        break
      } else if (arr[mid] > find) {
        end = mid - 1
      } else {
        start = mid + 1
      }
    }
    answer.push(check ? dic[arr[mid]] : 0)
  }
  return answer.join(' ')
}

console.log(solution())

File: baekjoon/BinarySearch/10816_숫자 카드2.py
import sys
from collections import Counter
input = sys.stdin.readline

n = int(input())
nArr = list(map(int, input().split(" ")))
m = int(input())
mArr = list(map(int, input().split(" ")))


def solution():
    # 중복 요소 개수 카운팅
    dic = dict(Counter(nArr))
    # 중복 제거
    arr = list(set(nArr))
    arr.sort()
    answer = []

    for find in mArr:
        start = 0
        end = len(arr)-1
        check = False
        while start <= end:
            mid = (start+end) // 2
            if arr[mid] == find:
                check = True
                break
            elif arr[mid] > find:
                end = mid - 1
            else:
                start = mid + 1
        answer.append(dic[arr[mid]] if check else 0)

    print(*answer)


solution()
